test/realm_test.dart

```diff
@@ -620,4 +640,96 @@ Future<void> main([List<String>? args]) async {
     // We should not be in transaction here
     expect(realm.isInTransaction, false);
   });
+
+  test('Transitive updates', () {
+    final realm = getRealm(Configuration.local([Friend.schema, Party.schema]));
+
+    var alice = Friend(1);
+    var bob = Friend(2, bestFriend: alice);
+
+    realm.write(() => realm.add(bob));
+
+    expect(bob.isManaged, true);
+    expect(alice.isManaged, true);
+
+    final aliceAgain = Friend(1, age: 50);
+    final bobAgain = Friend(2, age: 50, bestFriend: aliceAgain);
+    final party = Party(guests: [aliceAgain, bobAgain]);
+
+    realm.write(() => realm.add(party, update: true));
+
+    expect(bobAgain.isManaged, true);
+    expect(aliceAgain.isManaged, true);
+
+    expect(bob.age, 50);
+    expect(alice.age, 50);
+    expect(bob, bobAgain);
+    expect(alice, aliceAgain);
+    expect(bob.bestFriend, alice);
+  });
+
+  test('Cycles updated correctly', () {
+    final realm = getRealm(Configuration.local([Friend.schema]));
+
+    var alice = Friend(1);
+    var bob = Friend(2, bestFriend: alice);
+    alice.bestFriend = bob; // form cycle
+
+    realm.write(() => realm.add(alice));
+
+    final aliceAgain = Friend(1);
+    final bobAgain = Friend(2, bestFriend: aliceAgain);
+    aliceAgain.bestFriend = bobAgain;
+
+    alice = realm.write(() => realm.add(aliceAgain, update: true));
```

Comment:
What is being updated here? It seems like we're reinserting the exact same objects as before?

Reply:
This test would throw an exception until the update flag was correctly working, independent of testing against `car1` or `car2`. I can see why this test reads less naturally now, but it was added as part of a TDD flow. I can revisit the tests, but I risk changing them in the name of readability in a way that may not capture the original error anymore.

Anyway, I will do that, as I can see you have other comments to the same effect.